optimum/habana/transformers/generation/utils.py

```diff
@@ -199,19 +199,20 @@ def _prepare_decoder_input_ids_for_generation(
         # 2. `decoder_start_token_id` must have shape (batch_size, 1)
         if device is None:
             device = self.device
-        if token_idx is None:
-            if decoder_start_token_id.ndim == 1:
-                if decoder_start_token_id.shape[0] != batch_size:
-                    raise ValueError(
-                        f"`decoder_start_token_id` expected to have length {batch_size} but got {decoder_start_token_id.shape[0]}"
-                    )
-                decoder_start_token_id = decoder_start_token_id.view(-1, 1)
-            else:
-                decoder_start_token_id = (
-                    torch.ones((batch_size, 1), dtype=torch.long, device=device) * decoder_start_token_id
+        if decoder_start_token_id.ndim == 1:
+            if decoder_start_token_id.shape[0] != batch_size:
+                raise ValueError(
+                    f"`decoder_start_token_id` expected to have length {batch_size} but got {decoder_start_token_id.shape[0]}"
                 )
+            decoder_start_token_id = decoder_start_token_id.view(-1, 1)
         else:
-            # creating padded decoder_input_ids to achieve static shapes. Later new tokens once generated are copied in to decoder_input_ids based on token_idx
+            decoder_start_token_id = (
+                torch.ones((batch_size, 1), dtype=torch.long, device=device) * decoder_start_token_id
+            )
+
+        if token_idx is not None:
```

Comment:
@malkomes this is a bit confusing! Are you repeating the same ops again here? Why did you remove the if condition in the first place if you want to repeat the same operation again

```py
            decoder_start_token_id = (
                torch.ones((batch_size, 1), dtype=torch.long, device=device) * decoder_start_token_id
            )
```

Reply:
It should only have the  max_length and padding

Reply:
Well, here we are adding a logic to handle the case where we have a batch of tokens on the `decoder_start_token_id`. And, indeed. there different ways to implement this. In fact, in the other PR I did something a little bit "more cute" where I handle the `decoder_start_token_id` batch case and the static shape with more code changes using different variables names and etc. 

This time I realize that it's much easier to maintain the code if we keep the code similar to the [transformers code](https://github.com/huggingface/transformers/blob/1d063793318b20654ebb850f48f43e0a247ab7bb/src/transformers/generation/utils.py#L660-L669). It so happens that the max_length padding works for batch and non-batch if we do it afterwards. 

I was hoping that my comments here would explain it: https://github.com/malkomes/optimum-habana/pull/2 let me know if that helps or not. And maybe we should add notes on the code since it is a little bit confusing.

Reply:
thanks for the explanation, I agree for readability and maintainability, your separation makes sense!